templates/api_paginated_listers.mustache

```diff
@@ -0,0 +1,43 @@
+{{>partial_header}}
+package {{packageName}}
+
+{{#operations}}
+import (
+	"bytes"
+	"context"
+	"io"
+	"net/http"
+	"net/url"
+{{#imports}}	"{{import}}"
+{{/imports}}
+)
+
+{{#operation}}
+{{#vendorExtensions}}
+{{#x-paginated}}
```

Comment:
Is this vendor extension already supported or do we need to change the generator code to expose it to the templates?

Reply:
If referencing `x-paginated` does require a change to the generator, perhaps an alternative here would be 
to [use `go generate`](https://go.dev/blog/generate); we could use a custom `api.mustache` template that adds a `//go:generate` comment to every operation, regardless of parameters and the command in that generate comment could inspect the operation and add an `ExecuteWithPagination` function if the request object has a `Page` function.

Reply:
I think, in practice, go:generate works in applications (metal-cli) but not library packages (metal-go). 

The generated code would need to be generated and stored as an artifact in a ready-to-import home. We'd have a second step and a second directory.

`x-` vendor prefixes are part of the OAS spec: https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.0.0.md#specification-extensions

@t0mk added [a script](https://github.com/equinix-labs/metal-go/commit/5fc7ae48e72131551a2ac257005ccfc810441ee7#diff-b0fedfea2cf4fa0754f8f6b40435afbe2966376565f906baaba636fb7b9267e6) that finds `Page.yaml` dependent `GET` operations and adds `x-paginated: true`.

OpenAPI Generator supports `x-*` out of the box (https://openapi-generator.tech/docs/templating/#extensions) and makes them available as `vendorExtensions` (which isn't directly documented, but you can see that vendorExtensions is part of the model available to templates here: https://openapi-generator.tech/docs/templating/#models)

Reply:
I'm aware of vendor extensions as an OpenAPI standard, and that there is some support for them in openapi-generator, but I'm unclear on whether you can arbitrarily add vendor extensions to a spec and have openapi-generator expose those to the templates.  The page section you linked to says the following:
>The following are vendor extensions supported by OpenAPI Generator. The list may not be up-to-date, the best way is to look for "x-" in the built-in mustache templates.

Further up the page at the start of the templating docs, it says this:

> If you need to include additional files in your generated output, manipulate the OpenAPI document inputs, or implement your own vendor extensions or other logic, you'll want to read [customization](https://openapi-generator.tech/docs/customization) after you read this document.

Reply:
Does this work because it's a separate template?  Or can we reference arbitrary vendor extensions in a custom `api.mustache`?

Reply:
@ctreatma 

> > The following are vendor extensions supported by OpenAPI Generator. The list may not be up-to-date, the best way is to look for "x-" in the built-in mustache templates.

I think they mean that there are some existing extensions that people can use. If there's some functionality that oag devs consider generally useful (like enum descriptions https://openapi-generator.tech/docs/templating#enum), they define extension that users can add to their spec to use that functionality.

> I'm unclear on whether you can arbitrarily add vendor extensions to a spec and have openapi-generator expose those to the templates.

I spent more-than-I'd-like hours finding out if that's doable. This PR is a demonstration that it possible to add keys 
 in the `^x-` format to e.g. `paths['/projects/id/devices']['get']`, and have them exposed in the oag templates. They appear in `operation->vendorExtensions`. 

> Does this work because it's a separate template? Or can we reference arbitrary vendor extensions in a custom api.mustache?

There doesn't have to be a separate template, it could be added to `api.mustache`. I just figured it's better to put it away and don't taint the default template. We can do it in Go and have the new fucntions in the same namespace. In Python there's module per file IIUC, so it might be more complicated.

Reply:
Now I see that we don't want the listsers file generated for APIs that don't have any paginated method. So it seems we'll need to extend api.mustache :/. Because then we can just add the methods to existing source files.

Reply:
Yeah, I poked at it a bit more and realized the reason I couldn't reproduce the output locally was because I don't have `python`, so the pagination marks weren't making it into the input spec, not because of a limitation in the generator.